In [11]:
import pandas as pd
import numpy as np
import yfinance as yf

/usr/local/lib/python3.10/dist-packages/yfinance/base.py:48: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  _empty_series = pd.Series()


In [ ]:
# The Big Seven
company_list = ["TSLA", "AAPL", "GOOG", "AMZN", "MSFT", "META", "NVDA"]
company_names = ["TESLA", "APPLE", "MICROSOFT", "AMAZON", "MICROSOFT", "META", "NVIDIA"]

# Download stock data from yfinance
def download_stock(stock):
    df = yf.download(
        [stock],
        start="2010-01-01",
        end="2024-01-01",
        group_by="ticker",
        progress=False
    )
    df.round(4)
    df.insert(loc=0, column="Company", value=stock)
    return df

# Print the general information of the given stock
def info_stock(stock_name, stock_data):
    print(f'{stock_name}\'s Historical Data: \n\n {stock_data} \n\n')
    print(f'{stock_name}\'s Column Information: \n')
    print(stock_data.info(),'\n\n')
    print(f'{stock_name}\'s Statistic Summary: \n\n {stock_data.describe().round(4)}')
    print('-'*75)
    print()

In [ ]:
stock_dict = {}
for stock in company_list:
    stock_dict[stock] = download_stock(stock)

for stock_name, stock_data in stock_dict.items():
    info_stock(stock_name, stock_data)

In [14]:
company_list = ["TSLA", "AAPL", "GOOG", "AMZN", "MSFT", "META", "NVDA"]


In [15]:
stock_dict["TSLA"]

Company        Open        High         Low       Close  \
Date                                                                 
2010-06-29    TSLA    1.266667    1.666667    1.169333    1.592667   
2010-06-30    TSLA    1.719333    2.028000    1.553333    1.588667   
2010-07-01    TSLA    1.666667    1.728000    1.351333    1.464000   
2010-07-02    TSLA    1.533333    1.540000    1.247333    1.280000   
2010-07-06    TSLA    1.333333    1.333333    1.055333    1.074000   
...            ...         ...         ...         ...         ...   
2023-12-22    TSLA  256.760010  258.220001  251.369995  252.539993   
2023-12-26    TSLA  254.490005  257.970001  252.910004  256.609985   
2023-12-27    TSLA  258.350006  263.339996  257.519989  261.440002   
2023-12-28    TSLA  263.660004  265.130005  252.710007  253.179993   
2023-12-29    TSLA  255.100006  255.190002  247.429993  248.479996   

             Adj Close     Volume  
Date                               
2010-06-29    1.592667  281494500  
2010-06-30    1.588667  257806500  
2010-07-01    1.464000  123282000  
2010-07-02    1.280000   77097000  
2010-07-06    1.074000  103003500  
...                ...        ...  
2023-12-22  252.539993   93249800  
2023-12-26  256.609985   86892400  
2023-12-27  261.440002  106494400  
2023-12-28  253.179993  113619900  
2023-12-29  248.479996  100615300  

[3400 rows x 7 columns]

In [16]:
stock_dict["AAPL"]

Company        Open        High         Low       Close  \
Date                                                                 
2010-01-04    AAPL    7.622500    7.660714    7.585000    7.643214   
2010-01-05    AAPL    7.664286    7.699643    7.616071    7.656429   
2010-01-06    AAPL    7.656429    7.686786    7.526786    7.534643   
2010-01-07    AAPL    7.562500    7.571429    7.466071    7.520714   
2010-01-08    AAPL    7.510714    7.571429    7.466429    7.570714   
...            ...         ...         ...         ...         ...   
2023-12-22    AAPL  195.179993  195.410004  192.970001  193.600006   
2023-12-26    AAPL  193.610001  193.889999  192.830002  193.050003   
2023-12-27    AAPL  192.490005  193.500000  191.089996  193.149994   
2023-12-28    AAPL  194.139999  194.660004  193.169998  193.580002   
2023-12-29    AAPL  193.899994  194.399994  191.729996  192.529999   

             Adj Close     Volume  
Date                               
2010-01-04    6.470740  493729600  
2010-01-05    6.481929  601904800  
2010-01-06    6.378825  552160000  
2010-01-07    6.367033  477131200  
2010-01-08    6.409361  447610800  
...                ...        ...  
2023-12-22  193.353287   37122800  
2023-12-26  192.803986   28919300  
2023-12-27  192.903839   48087700  
2023-12-28  193.333298   34049900  
2023-12-29  192.284637   42628800  

[3522 rows x 7 columns]

In [17]:
stock_dict["GOOG"]

Company        Open        High         Low       Close  \
Date                                                                 
2010-01-04    GOOG   15.615220   15.678981   15.547723   15.610239   
2010-01-05    GOOG   15.620949   15.637387   15.480475   15.541497   
2010-01-06    GOOG   15.588072   15.588072   15.102393   15.149715   
2010-01-07    GOOG   15.178109   15.193053   14.760922   14.797037   
2010-01-08    GOOG   14.744733   15.024933   14.672753   14.994298   
...            ...         ...         ...         ...         ...   
2023-12-22    GOOG  142.130005  143.250000  142.054993  142.720001   
2023-12-26    GOOG  142.979996  143.945007  142.500000  142.820007   
2023-12-27    GOOG  142.830002  143.320007  141.050995  141.440002   
2023-12-28    GOOG  141.850006  142.270004  140.828003  141.279999   
2023-12-29    GOOG  140.679993  141.434998  139.899994  140.929993   

             Adj Close     Volume  
Date                               
2010-01-04   15.610239   78541293  
2010-01-05   15.541497  120638494  
2010-01-06   15.149715  159744526  
2010-01-07   14.797037  257533695  
2010-01-08   14.994298  189680313  
...                ...        ...  
2023-12-22  142.720001   18494700  
2023-12-26  142.820007   11170100  
2023-12-27  141.440002   17288400  
2023-12-28  141.279999   12192500  
2023-12-29  140.929993   14872700  

[3522 rows x 7 columns]

In [18]:
stock_dict["AMZN"]

Company        Open        High         Low       Close  \
Date                                                                 
2010-01-04    AMZN    6.812500    6.830500    6.657000    6.695000   
2010-01-05    AMZN    6.671500    6.774000    6.590500    6.734500   
2010-01-06    AMZN    6.730000    6.736500    6.582500    6.612500   
2010-01-07    AMZN    6.600500    6.616000    6.440000    6.500000   
2010-01-08    AMZN    6.528000    6.684000    6.451500    6.676000   
...            ...         ...         ...         ...         ...   
2023-12-22    AMZN  153.770004  154.350006  152.710007  153.419998   
2023-12-26    AMZN  153.559998  153.979996  153.029999  153.410004   
2023-12-27    AMZN  153.559998  154.779999  153.119995  153.339996   
2023-12-28    AMZN  153.720001  154.080002  152.949997  153.380005   
2023-12-29    AMZN  153.100006  153.889999  151.029999  151.940002   

             Adj Close     Volume  
Date                               
2010-01-04    6.695000  151998000  
2010-01-05    6.734500  177038000  
2010-01-06    6.612500  143576000  
2010-01-07    6.500000  220604000  
2010-01-08    6.676000  196610000  
...                ...        ...  
2023-12-22  153.419998   29480100  
2023-12-26  153.410004   25067200  
2023-12-27  153.339996   31434700  
2023-12-28  153.380005   27057000  
2023-12-29  151.940002   39789000  

[3522 rows x 7 columns]

In [19]:
stock_dict["MSFT"]

Company        Open        High         Low       Close  \
Date                                                                 
2010-01-04    MSFT   30.620001   31.100000   30.590000   30.950001   
2010-01-05    MSFT   30.850000   31.100000   30.639999   30.959999   
2010-01-06    MSFT   30.879999   31.080000   30.520000   30.770000   
2010-01-07    MSFT   30.629999   30.700001   30.190001   30.450001   
2010-01-08    MSFT   30.280001   30.879999   30.240000   30.660000   
...            ...         ...         ...         ...         ...   
2023-12-22    MSFT  373.679993  375.179993  372.709991  374.579987   
2023-12-26    MSFT  375.000000  376.940002  373.500000  374.660004   
2023-12-27    MSFT  373.690002  375.059998  372.809998  374.070007   
2023-12-28    MSFT  375.369995  376.459991  374.160004  375.279999   
2023-12-29    MSFT  376.000000  377.160004  373.480011  376.040009   

             Adj Close    Volume  
Date                              
2010-01-04   23.431583  38409100  
2010-01-05   23.439157  49749600  
2010-01-06   23.295319  58182400  
2010-01-07   23.053047  50559700  
2010-01-08   23.212034  51197400  
...                ...       ...  
2023-12-22  373.888580  17091100  
2023-12-26  373.968445  12673100  
2023-12-27  373.379547  14905400  
2023-12-28  374.587280  14327000  
2023-12-29  375.345886  18723000  

[3522 rows x 7 columns]

In [20]:
stock_dict["META"]

Company        Open        High         Low       Close  \
Date                                                                 
2012-05-18    META   42.049999   45.000000   38.000000   38.230000   
2012-05-21    META   36.529999   36.660000   33.000000   34.029999   
2012-05-22    META   32.610001   33.590000   30.940001   31.000000   
2012-05-23    META   31.370001   32.500000   31.360001   32.000000   
2012-05-24    META   32.950001   33.209999   31.770000   33.029999   
...            ...         ...         ...         ...         ...   
2023-12-22    META  355.579987  357.200012  351.220001  353.390015   
2023-12-26    META  354.989990  356.980011  353.450012  354.829987   
2023-12-27    META  356.070007  359.000000  355.309998  357.829987   
2023-12-28    META  359.700012  361.899994  357.809998  358.320007   
2023-12-29    META  358.989990  360.000000  351.820007  353.959991   

             Adj Close     Volume  
Date                               
2012-05-18   38.230000  573576400  
2012-05-21   34.029999  168192700  
2012-05-22   31.000000  101786600  
2012-05-23   32.000000   73600000  
2012-05-24   33.029999   50237200  
...                ...        ...  
2023-12-22  353.390015   11764200  
2023-12-26  354.829987    9898600  
2023-12-27  357.829987   13207900  
2023-12-28  358.320007   11798800  
2023-12-29  353.959991   14980500  

[2923 rows x 7 columns]

In [21]:
stock_dict["NVDA"]

Company        Open        High         Low       Close  \
Date                                                                 
2010-01-04    NVDA    4.627500    4.655000    4.527500    4.622500   
2010-01-05    NVDA    4.605000    4.740000    4.605000    4.690000   
2010-01-06    NVDA    4.687500    4.730000    4.642500    4.720000   
2010-01-07    NVDA    4.695000    4.715000    4.592500    4.627500   
2010-01-08    NVDA    4.590000    4.670000    4.562500    4.637500   
...            ...         ...         ...         ...         ...   
2023-12-22    NVDA  491.950012  493.829987  484.670013  488.299988   
2023-12-26    NVDA  489.679993  496.000000  489.600006  492.790009   
2023-12-27    NVDA  495.109985  496.799988  490.850006  494.170013   
2023-12-28    NVDA  496.429993  498.839996  494.119995  495.220001   
2023-12-29    NVDA  498.130005  499.970001  487.510010  495.220001   

             Adj Close    Volume  
Date                              
2010-01-04    4.240430  80020400  
2010-01-05    4.302351  72864800  
2010-01-06    4.329870  64916800  
2010-01-07    4.245015  54779200  
2010-01-08    4.254190  47816800  
...                ...       ...  
2023-12-22  488.299988  25213900  
2023-12-26  492.790009  24420000  
2023-12-27  494.170013  23364800  
2023-12-28  495.220001  24658700  
2023-12-29  495.220001  38869000  

[3522 rows x 7 columns]

In [24]:
import numpy as np
import pandas as pd
  # tqdm'i import et
from tqdm import tqdm
# stock_dict sözlüğünüz varsayılan olarak mevcut ve her bir hisse senedi için DataFrame içeriyor.
# Örnek olarak, sadece bir hisse senedi için (NVDA) nasıl yapılacağını gösteriyorum.
# Gerçek senaryoda, tüm hisse senetleri için bu işlemi yapmalısınız.

# Tüm hisse senetleri için günlük getirileri birleştirin
# combined_returns DataFrame'inin her sütunu bir hisse senedinin günlük getirilerini içerecek şekilde olmalıdır.
combined_returns = pd.DataFrame()
for stock in stock_dict.keys():
    combined_returns[stock] = stock_dict[stock]['Close'].pct_change().dropna()

# Genetik algoritma parametreleri
population_size = 1000
num_generations = 90
mutation_rate = 0.10
num_assets = len(stock_dict)


population = np.random.dirichlet(np.ones(num_assets), size=population_size)


def fitness(weights, returns):
    # Portföy getirisini ve volatilitesini hesapla
    port_return = np.sum(weights * returns.mean()) * 252 #closed dan günlük kar 252 gün var yılda
    port_volatility = np.sqrt(np.dot(weights.T, np.dot(returns.cov() * 252, weights))) # risk volatile göre
    # Sharpe oranını hesapla
    sharpe_ratio = port_return / port_volatility
    return -sharpe_ratio  # maksimize etmek için negatif değer


for generation in tqdm(range(num_generations), desc="Evolving Generations"):
    fitness_scores = np.array([fitness(individual, combined_returns) for individual in population])
    best_indices = np.argsort(fitness_scores)[:population_size // 2]
    survivors = population[best_indices]

    new_population = []
    while len(new_population) < population_size:
        parents = np.random.choice(survivors.shape[0], 2, replace=False)
        cross_point = np.random.randint(1, num_assets)
        child = np.concatenate([survivors[parents[0], :cross_point], survivors[parents[1], cross_point:]])
        if np.random.rand() < mutation_rate:
            mutation_point = np.random.randint(num_assets)
            child[mutation_point] = np.random.rand()
            child /= np.sum(child)  # ağırlıkları normalize et
        new_population.append(child)
    population = np.array(new_population)

best_index = np.argmin([fitness(individual, combined_returns) for individual in population])
best_weights = population[best_index]
port_volatility = np.sqrt(np.dot(best_weights.T, np.dot(combined_returns.cov() * 252, best_weights)))

print("Optimized Portfolio Volatility (Risk):", port_volatility)
print("Optimized Portfolio Weights:", best_weights)

Evolving Generations: 100%|██████████| 90/90 [03:52<00:00,  2.59s/it]


Optimized Portfolio Volatility (Risk): 0.29778302145092483
Optimized Portfolio Weights: [0.19663248 0.23689857 0.00045083 0.19400797 0.07716648 0.00060798
 0.29423568]


In [25]:
population

array([[0.21699566, 0.26191447, 0.00049752, ..., 0.08515781, 0.00067094,
        0.32470661],
       [0.21699566, 0.26191447, 0.00049752, ..., 0.08515781, 0.00067094,
        0.32470661],
       [0.21451684, 0.25892251, 0.00049184, ..., 0.08418502, 0.00066328,
        0.2295669 ],
       ...,
       [0.21699566, 0.26191447, 0.00049752, ..., 0.08515781, 0.00067094,
        0.32470661],
       [0.21699566, 0.26191447, 0.00049752, ..., 0.08515781, 0.00067094,
        0.32470661],
       [0.21699566, 0.26191447, 0.00049752, ..., 0.08515781, 0.00067094,
        0.32470661]])

elitizm var en iyi genler devam ediyor.
100 popülation 90 gen %10 mutation
Optimized Portfolio Volatility (Risk): 0.311111318818031
Optimized Portfolio Weights: [0.20175045 0.2595938  0.00259982 0.18026265 0.10944589 0.00162085
 0.3026919 ]

 Optimized Portfolio Volatility (Risk): 0.3085001391434136
Optimized Portfolio Weights: [0.20043154 0.2430891  0.01426032 0.19111831 0.08346771 0.00763725
 0.30342829]

1000 popülation 90 gen %10 mutation
 Optimized Portfolio Volatility (Risk): 0.29778302145092483
Optimized Portfolio Weights: [0.19663248 0.23689857 0.00045083 0.19400797 0.07716648 0.00060798
 0.29423568]

In [30]:
#daha çeşitlendirmeye çalışıyoruz

import numpy as np
import pandas as pd

# stock_dict sözlüğünüz varsayılan olarak mevcut ve her bir hisse senedi için DataFrame içeriyor.
# Örnek olarak, sadece bir hisse senedi için (NVDA) nasıl yapılacağını gösteriyorum.
# Gerçek senaryoda, tüm hisse senetleri için bu işlemi yapmalısınız.

# Tüm hisse senetleri için günlük getirileri birleştirin
# combined_returns DataFrame'inin her sütunu bir hisse senedinin günlük getirilerini içerecek şekilde olmalıdır.
combined_returns = pd.DataFrame()
for stock in stock_dict.keys():
    combined_returns[stock] = stock_dict[stock]['Close'].pct_change().dropna()

# Genetik algoritma parametreleri
population_size = 1000
num_generations = 90
mutation_rate = 0.20
num_assets = len(stock_dict)


population = np.random.dirichlet(np.ones(num_assets)* 0.5, size=population_size) # * 0.5 eklendi


def fitness(weights, returns):
    # Portföy getirisini ve volatilitesini hesapla
    port_return = np.sum(weights * returns.mean()) * 252 #closed dan günlük kar
    port_volatility = np.sqrt(np.dot(weights.T, np.dot(returns.cov() * 252, weights))) # risk volatile göre
    # Sharpe oranını hesapla
    sharpe_ratio = port_return / port_volatility
    return -sharpe_ratio  # maksimize etmek için negatif değer


for generation in tqdm(range(num_generations), desc="Evolving Generations"):
    fitness_scores = np.array([fitness(individual, combined_returns) for individual in population])
    best_indices = np.argsort(fitness_scores)[:population_size // 2]
    survivors = population[best_indices]

    new_population = []
    while len(new_population) < population_size:
        parents = np.random.choice(survivors.shape[0], 2, replace=False)
        cross_points = np.sort(np.random.choice(np.arange(1, num_assets - 1), 2, replace=False))
        child = np.concatenate([survivors[parents[0], :cross_points[0]],
                                survivors[parents[1], cross_points[0]:cross_points[1]], # more cross
                                survivors[parents[0], cross_points[1]:]])

        # Mutasyon mekanizmasını güçlendir
        if np.random.rand() < mutation_rate:
            mutation_point = np.random.randint(num_assets)
            child += np.random.normal(0, 0.1, num_assets)  # Normal dağılımlı mutasyon
            child[mutation_point] = np.random.rand()
            child = np.clip(child, 0, 1)  # Değerleri [0, 1] aralığında sınırla
            child /= np.sum(child)  # Ağırlıkları normalize et

        new_population.append(child)
    population = np.array(new_population)

best_index = np.argmin([fitness(individual, combined_returns) for individual in population])
best_weights = population[best_index]
port_volatility = np.sqrt(np.dot(best_weights.T, np.dot(combined_returns.cov() * 252, best_weights)))

print("Optimized Portfolio Volatility (Risk):", port_volatility)
print("Optimized Portfolio Weights:", best_weights)

Evolving Generations: 100%|██████████| 90/90 [03:53<00:00,  2.59s/it]


Optimized Portfolio Volatility (Risk): 0.30520168333963776
Optimized Portfolio Weights: [0.20239053 0.23951137 0.         0.19885441 0.08306514 0.
 0.30101917]


In [31]:
population


array([[0.20239053, 0.23951137, 0.        , ..., 0.08306514, 0.        ,
        0.30101917],
       [0.20239053, 0.23951137, 0.        , ..., 0.08306514, 0.        ,
        0.30101917],
       [0.20239053, 0.23951137, 0.        , ..., 0.08306514, 0.        ,
        0.30101917],
       ...,
       [0.20239053, 0.23951137, 0.        , ..., 0.08306514, 0.        ,
        0.30101917],
       [0.10947619, 0.48472287, 0.08283252, ..., 0.12818311, 0.03694219,
        0.07173891],
       [0.20239053, 0.23951137, 0.        , ..., 0.08306514, 0.        ,
        0.30101917]])

more random popülasyon 100 gen 90 mut %20


Optimized Portfolio Volatility (Risk): 0.287117900813033
Optimized Portfolio Weights: [0.19415363 0.22200493 0.         0.18048821 0.05643044 0.
 0.2954313 ]

 Optimized Portfolio Volatility (Risk): 0.32968149054507734
Optimized Portfolio Weights: [0.23015095 0.2791906  0.         0.21386455 0.0907295  0.
 0.30104682]
 popülasyon 1000 gen 90 mut %20

 Optimized Portfolio Volatility (Risk): 0.30520168333963776
Optimized Portfolio Weights: [0.20239053 0.23951137 0.         0.19885441 0.08306514 0.
 0.30101917]